cli.go

```diff
@@ -0,0 +1,502 @@
+	// handle ctrl-c while waiting for the callback
+	ch := make(chan os.Signal, 1)
+	signal.Notify(ch, os.Interrupt)
+	defer signal.Stop(ch)
+
+	doneCh := make(chan loginResp)
+
+	mount, ok := m["mount"]
+	if !ok {
+		mount = defaultMount
+	}
+
+	port, ok := m["port"]
+	if !ok {
+		port = defaultPort
+	}
+
+	role := m["role"]
+	if role == "" {
+		return nil, errors.New("a 'role' must be specified")
+	}
+
+	secret, err := fetchAuthURL(c, role, mount, port)
+	if err != nil {
+		return nil, err
+	}
+
+	authURL := secret.Data["auth_url"].(string)
+	if authURL == "" {
+		return nil, errors.New(fmt.Sprintf("Unable to authorize role %q. Check Vault logs for more information.", role))
+	}
+
+	fmt.Fprintf(os.Stderr, "Complete the login via your OIDC provider. Launching browser to:\n\n    %s\n\n\n", authURL)
+	if err := openURL(authURL); err != nil {
+		fmt.Fprintf(os.Stderr, "Error attempting to automatically open browser: '%s'.\nPlease visit the authorization URL manually.", err)
+	}
+
+	// Set up callback handler
+	http.HandleFunc(fmt.Sprintf("/v1/auth/%s/oidc/callback", mount), func(w http.ResponseWriter, req *http.Request) {
+		var response string
+
+		query := req.URL.Query()
+		code := query.Get("code")
+		state := query.Get("state")
+		data := map[string][]string{
+			"code":  {code},
+			"state": {state},
+		}
+
+		secret, err := c.Logical().ReadWithData(fmt.Sprintf("auth/%s/oidc/callback", mount), data)
+		if err != nil {
+			summary, detail := parseError(err)
+			response = errorHTML(summary, detail)
+		} else {
+			response = successHTML
+		}
+
+		w.Write([]byte(response))
+		doneCh <- loginResp{secret, err}
+	})
+
+	// Start local server
+	go func() {
+		if err := http.ListenAndServe(":"+port, nil); err != nil && err != http.ErrServerClosed {
+			fmt.Fprintf(os.Stderr, "Error listening for callback: %v\n\n", err.Error())
+		}
+	}()
+
+	// Wait for either the callback to finish or SIGINT to be received
+	select {
+	case s := <-doneCh:
+		return s.secret, s.err
+	case <-ch:
+		return nil, errors.New("interrupted")
+	}
+}
+
+func fetchAuthURL(c *api.Client, role, mount, port string) (*api.Secret, error) {
+	data := map[string]interface{}{
+		"role":         role,
+		"redirect_uri": fmt.Sprintf("http://localhost:%s/v1/auth/%s/oidc/callback", port, mount),
+	}
+
+	return c.Logical().Write(fmt.Sprintf("auth/%s/oidc/auth_url", mount), data)
+}
+
+// openURL opens the specified URL in the default browser of the user.
+// Source: https://stackoverflow.com/a/39324149/453290
+func openURL(url string) error {
+	var cmd string
+	var args []string
+
+	switch runtime.GOOS {
+	case "windows":
+		cmd = "cmd"
+		args = []string{"/c", "start"}
+	case "darwin":
+		cmd = "open"
+	default: // "linux", "freebsd", "openbsd", "netbsd"
+		cmd = "xdg-open"
+	}
+	args = append(args, url)
+	return exec.Command(cmd, args...).Start()
+}
+
+// parseError converts error from the API into summary and detailed portions.
```

Comment:
Would it be possible here to leave in the comment an example of the error before and after?

Reply:
Yep